File: jung-visualization/src/main/java/edu/uci/ics/jung/visualization/RenderContext.java
package edu.uci.ics.jung.visualization;

import com.google.common.graph.Network;
import edu.uci.ics.jung.visualization.layout.NetworkElementAccessor;
import edu.uci.ics.jung.visualization.picking.PickedState;
import edu.uci.ics.jung.visualization.renderers.EdgeLabelRenderer;
import edu.uci.ics.jung.visualization.renderers.NodeLabelRenderer;
import edu.uci.ics.jung.visualization.transform.shape.GraphicsDecorator;
import edu.uci.ics.jung.visualization.util.Context;
import edu.uci.ics.jung.visualization.util.EdgeIndexFunction;
import java.awt.BasicStroke;
import java.awt.Font;
import java.awt.Paint;
import java.awt.Shape;
import java.awt.Stroke;
import java.util.function.Function;
import java.util.function.Predicate;
import javax.swing.CellRendererPane;
import javax.swing.Icon;
import javax.swing.JComponent;

public interface RenderContext<N, E> {

  float[] dotting = {1.0f, 3.0f};
  float[] dashing = {5.0f};

  /**
   * A stroke for a dotted line: 1 pixel width, round caps, round joins, and an array of {1.0f,
   * 3.0f}.
   */
  Stroke DOTTED =
      new BasicStroke(1.0f, BasicStroke.CAP_ROUND, BasicStroke.JOIN_ROUND, 1.0f, dotting, 0f);

  /**
   * A stroke for a dashed line: 1 pixel width, square caps, beveled joins, and an array of {5.0f}.
   */
  Stroke DASHED =
      new BasicStroke(1.0f, BasicStroke.CAP_SQUARE, BasicStroke.JOIN_BEVEL, 1.0f, dashing, 0f);

  /** Specifies the offset for the edge labels. */
  int LABEL_OFFSET = 10;

  int getLabelOffset();

  void setLabelOffset(int labelOffset);

  float getArrowPlacementTolerance();

  void setArrowPlacementTolerance(float arrow_placement_tolerance);

  Shape getEdgeArrow();

  void setEdgeArrow(Shape shape);

  boolean renderEdgeArrow();

  void setRenderEdgeArrow(boolean render);

  Function<? super E, Font> getEdgeFontFunction();

  void setEdgeFontFunction(Function<? super E, Font> edgeFontFunction);

  Predicate<E> getEdgeIncludePredicate();

  void setEdgeIncludePredicate(Predicate<E> edgeIncludePredicate);

  public float getEdgeLabelCloseness();

  public void setEdgeLabelCloseness(float closeness);

  EdgeLabelRenderer getEdgeLabelRenderer();

  void setEdgeLabelRenderer(EdgeLabelRenderer edgeLabelRenderer);

  Function<? super E, Paint> getEdgeFillPaintFunction();

  void setEdgeFillPaintFunction(Function<? super E, Paint> edgePaintFunction);

  Function<? super E, Paint> getEdgeDrawPaintFunction();

  void setEdgeDrawPaintFunction(Function<? super E, Paint> edgeDrawPaintFunction);

  Function<? super E, Paint> getArrowDrawPaintFunction();

  void setArrowDrawPaintFunction(Function<? super E, Paint> arrowDrawPaintFunction);

  Function<? super E, Paint> getArrowFillPaintFunction();

  void setArrowFillPaintFunction(Function<? super E, Paint> arrowFillPaintFunction);

  Function<Context<Network<N, E>, E>, Shape> getEdgeShapeFunction();

  void setEdgeShapeFunction(Function<Context<Network<N, E>, E>, Shape> edgeShapeFunction);

  Function<? super E, String> getEdgeLabelFunction();

  void setEdgeLabelFunction(Function<? super E, String> edgeStringer);

  Function<? super E, Stroke> edgeStrokeFunction();

  void setEdgeStrokeFunction(Function<? super E, Stroke> edgeStrokeFunction);

  Function<? super E, Stroke> getEdgeArrowStrokeFunction();

  void setEdgeArrowStrokeFunction(Function<? super E, Stroke> edgeArrowStrokeFunction);

  GraphicsDecorator getGraphicsContext();

  void setGraphicsContext(GraphicsDecorator graphicsContext);

  EdgeIndexFunction<N, E> getParallelEdgeIndexFunction();

  void setParallelEdgeIndexFunction(EdgeIndexFunction<N, E> parallelEdgeIndexFunction);

  PickedState<E> getPickedEdgeState();

  void setPickedEdgeState(PickedState<E> pickedEdgeState);

  PickedState<N> getPickedNodeState();

  void setPickedNodeState(PickedState<N> pickedNodeState);

  CellRendererPane getRendererPane();

  void setRendererPane(CellRendererPane rendererPane);

  JComponent getScreenDevice();

  void setScreenDevice(JComponent screenDevice);

  Function<? super N, Font> getNodeFontFunction();

  void setNodeFontFunction(Function<? super N, Font> nodeFontFunction);

  Function<N, Icon> getNodeIconFunction();

  void setNodeIconFunction(Function<N, Icon> nodeIconFunction);

  Predicate<N> getNodeIncludePredicate();

  void setNodeIncludePredicate(Predicate<N> nodeIncludePredicate);

  NodeLabelRenderer getNodeLabelRenderer();

  void setNodeLabelRenderer(NodeLabelRenderer nodeLabelRenderer);

  Function<? super N, Paint> getNodeFillPaintFunction();

  void setNodeFillPaintFunction(Function<? super N, Paint> nodeFillPaintFunction);

  Function<? super N, Paint> getNodeDrawPaintFunction();

  void setNodeDrawPaintFunction(Function<? super N, Paint> nodeDrawPaintFunction);

  Function<? super N, Shape> getNodeShapeFunction();

  void setNodeShapeFunction(Function<? super N, Shape> nodeShapeFunction);

  Function<? super N, String> getNodeLabelFunction();

  void setNodeLabelFunction(Function<? super N, String> nodeStringer);

  Function<? super N, Paint> getNodeLabelDrawPaintFunction();

  void setNodeLabelDrawPaintFunction(Function<? super N, Paint> nodeLabelDrawPaintFunction);

  Function<? super N, Stroke> getNodeStrokeFunction();

  void setNodeStrokeFunction(Function<? super N, Stroke> nodeStrokeFunction);

  class DirectedEdgeArrowPredicate implements Predicate<Network<?, ?>> {

    public boolean test(Network<?, ?> graph) {
      return graph.isDirected();
    }
  }

  class UndirectedEdgeArrowPredicate implements Predicate<Network<?, ?>> {

    public boolean test(Network<?, ?> graph) {
      return !graph.isDirected();
    }
  }

  MultiLayerTransformer getMultiLayerTransformer();

  void setMultiLayerTransformer(MultiLayerTransformer basicTransformer);

  /** @return the pickSupport */
  NetworkElementAccessor<N, E> getPickSupport();

  /** @param pickSupport the pickSupport to set */
  void setPickSupport(NetworkElementAccessor<N, E> pickSupport);
}

File: jung-visualization/src/main/kotlin/edu/uci/ics/jung/visualization/RenderContext.kt
package edu.uci.ics.jung.visualization

import com.google.common.graph.Network
import edu.uci.ics.jung.visualization.layout.NetworkElementAccessor
import edu.uci.ics.jung.visualization.picking.PickedState
import edu.uci.ics.jung.visualization.renderers.EdgeLabelRenderer
import edu.uci.ics.jung.visualization.renderers.NodeLabelRenderer
import edu.uci.ics.jung.visualization.transform.shape.GraphicsDecorator
import edu.uci.ics.jung.visualization.util.Context
import edu.uci.ics.jung.visualization.util.EdgeIndexFunction
import java.awt.BasicStroke
import java.awt.Font
import java.awt.Paint
import java.awt.Shape
import java.awt.Stroke
import java.util.function.Function
import java.util.function.Predicate
import javax.swing.CellRendererPane
import javax.swing.Icon
import javax.swing.JComponent

interface RenderContext<N : Any, E : Any> {

    fun getLabelOffset(): Int

    fun setLabelOffset(labelOffset: Int)

    fun getArrowPlacementTolerance(): Float

    fun setArrowPlacementTolerance(arrow_placement_tolerance: Float)

    fun getEdgeArrow(): Shape

    fun setEdgeArrow(shape: Shape)

    fun renderEdgeArrow(): Boolean

    fun setRenderEdgeArrow(render: Boolean)

    fun getEdgeFontFunction(): Function<in E, Font>

    fun setEdgeFontFunction(edgeFontFunction: Function<in E, Font>)

    fun getEdgeIncludePredicate(): Predicate<E>

    fun setEdgeIncludePredicate(edgeIncludePredicate: Predicate<E>)

    fun getEdgeLabelCloseness(): Float

    fun setEdgeLabelCloseness(closeness: Float)

    fun getEdgeLabelRenderer(): EdgeLabelRenderer

    fun setEdgeLabelRenderer(edgeLabelRenderer: EdgeLabelRenderer)

    fun getEdgeFillPaintFunction(): Function<in E, Paint>

    fun setEdgeFillPaintFunction(edgePaintFunction: Function<in E, Paint>)

    fun getEdgeDrawPaintFunction(): Function<in E, Paint>

    fun setEdgeDrawPaintFunction(edgeDrawPaintFunction: Function<in E, Paint>)

    fun getArrowDrawPaintFunction(): Function<in E, Paint>

    fun setArrowDrawPaintFunction(arrowDrawPaintFunction: Function<in E, Paint>)

    fun getArrowFillPaintFunction(): Function<in E, Paint>

    fun setArrowFillPaintFunction(arrowFillPaintFunction: Function<in E, Paint>)

    fun getEdgeShapeFunction(): Function<Context<Network<N, E>, E>, Shape>

    fun setEdgeShapeFunction(edgeShapeFunction: Function<Context<Network<N, E>, E>, Shape>)

    fun getEdgeLabelFunction(): Function<in E, String>

    fun setEdgeLabelFunction(edgeStringer: Function<in E, String>)

    fun edgeStrokeFunction(): Function<in E, Stroke>

    fun setEdgeStrokeFunction(edgeStrokeFunction: Function<in E, Stroke>)

    fun getEdgeArrowStrokeFunction(): Function<in E, Stroke>

    fun setEdgeArrowStrokeFunction(edgeArrowStrokeFunction: Function<in E, Stroke>)

    fun getGraphicsContext(): GraphicsDecorator?

    fun setGraphicsContext(graphicsContext: GraphicsDecorator)

    fun getParallelEdgeIndexFunction(): EdgeIndexFunction<N, E>

    fun setParallelEdgeIndexFunction(parallelEdgeIndexFunction: EdgeIndexFunction<N, E>)

    fun getPickedEdgeState(): PickedState<E>

    fun setPickedEdgeState(pickedEdgeState: PickedState<E>)

    fun getPickedNodeState(): PickedState<N>

    fun setPickedNodeState(pickedNodeState: PickedState<N>)

    fun getRendererPane(): CellRendererPane

    fun setRendererPane(rendererPane: CellRendererPane)

    fun getScreenDevice(): JComponent

    fun setScreenDevice(screenDevice: JComponent)

    fun getNodeFontFunction(): Function<in N, Font>

    fun setNodeFontFunction(nodeFontFunction: Function<in N, Font>)

    fun getNodeIconFunction(): Function<N, Icon>

    fun setNodeIconFunction(nodeIconFunction: Function<N, Icon>)

    fun getNodeIncludePredicate(): Predicate<N>

    fun setNodeIncludePredicate(nodeIncludePredicate: Predicate<N>)

    fun getNodeLabelRenderer(): NodeLabelRenderer

    fun setNodeLabelRenderer(nodeLabelRenderer: NodeLabelRenderer)

    fun getNodeFillPaintFunction(): Function<in N, Paint>

    fun setNodeFillPaintFunction(nodeFillPaintFunction: Function<in N, Paint>)

    fun getNodeDrawPaintFunction(): Function<in N, Paint>

    fun setNodeDrawPaintFunction(nodeDrawPaintFunction: Function<in N, Paint>)

    fun getNodeShapeFunction(): Function<in N, Shape>

    fun setNodeShapeFunction(nodeShapeFunction: Function<in N, Shape>)

    fun getNodeLabelFunction(): Function<in N, String>

    fun setNodeLabelFunction(nodeStringer: Function<in N, String>)

    fun getNodeLabelDrawPaintFunction(): Function<in N, Paint>

    fun setNodeLabelDrawPaintFunction(nodeLabelDrawPaintFunction: Function<in N, Paint>)

    fun getNodeStrokeFunction(): Function<in N, Stroke>

    fun setNodeStrokeFunction(nodeStrokeFunction: Function<in N, Stroke>)

    class DirectedEdgeArrowPredicate : Predicate<Network<*, *>> {
        override fun test(graph: Network<*, *>): Boolean = graph.isDirected
    }

    class UndirectedEdgeArrowPredicate : Predicate<Network<*, *>> {
        override fun test(graph: Network<*, *>): Boolean = !graph.isDirected
    }

    fun getMultiLayerTransformer(): MultiLayerTransformer

    fun setMultiLayerTransformer(basicTransformer: MultiLayerTransformer)

    /**
     * @return the pickSupport
     */
    fun getPickSupport(): NetworkElementAccessor<N, E>

    /**
     * @param pickSupport the pickSupport to set
     */
    fun setPickSupport(pickSupport: NetworkElementAccessor<N, E>)

    companion object {
        val dotting = floatArrayOf(1.0f, 3.0f)
        val dashing = floatArrayOf(5.0f)

        /**
         * A stroke for a dotted line: 1 pixel width, round caps, round joins, and an array of {1.0f,
         * 3.0f}.
         */
        val DOTTED: Stroke =
            BasicStroke(1.0f, BasicStroke.CAP_ROUND, BasicStroke.JOIN_ROUND, 1.0f, dotting, 0f)

        /**
         * A stroke for a dashed line: 1 pixel width, square caps, beveled joins, and an array of {5.0f}.
         */
        val DASHED: Stroke =
            BasicStroke(1.0f, BasicStroke.CAP_SQUARE, BasicStroke.JOIN_BEVEL, 1.0f, dashing, 0f)

        /** Specifies the offset for the edge labels. */
        const val LABEL_OFFSET: Int = 10
    }
}
